utils/lib/Component.ts

```diff
@@ -0,0 +1,51 @@
+import * as path from 'path';
+import * as fs from 'fs';
+import * as yaml from 'js-yaml';
+
+abstract class Component<ConfigType, MutationVariablesType> {
+  public localPath: string; // Local path to the component. Ex: python/flask
+  public configPath: string; // Absolute path to the config file within the repository
+  public config: ConfigType;
+  public isValid = true;
+  public basePath: string;
+
+  constructor(
+    localPath: string,
+    basePath: string = path.join(__dirname, '../..')
+  ) {
+    this.basePath = basePath;
+    this.localPath = localPath;
+    this.configPath = this.getConfigFilePath();
+    this.config = this.getConfigContent();
+  }
+
+  abstract getConfigFilePath(): string;
+  abstract getConfigContent(): ConfigType;
+  abstract getMutationVariables(): MutationVariablesType;
+
+  get fullPath() {
+    return path.join(this.basePath, this.configPath);
+  }
+
+  /**
+   * Helper method for reading YAML-based configuration files.
+   */
+  protected _getYamlConfigContent(): ConfigType {
+    if (!this.isValid) {
+      return this.config;
+    }
```

Comment:
This is a tad misleading. If the component isn't valid, then returning `this.config` will return `undefined` here. Perhaps we should update the generic to be `ConfigType | undefined`.